File: Algorithm/python_source/charpter11_bitmap/level2/count_bits.py
# coding=utf-8

#  方法1
def countBits1(num):
    bits = [0] * (num + 1)
    for i in range(num + 1):
        for j in range(32):
            bits[i] += (i >> j) & 1
    return bits


# 方法2
def countOnes(x):
    ones = 0
    while x > 0:
        x &= (x - 1)
        ones += 1
    return ones


def countBits(num):
    bits = []
    for i in range(num + 1):
        bits.append(countOnes(i))
    return bits


if __name__ == '__main__':
    print(countBits1(3))

File: Algorithm/Java/src/main/java/com/yugutou/charpter11_bit/level2/CountBits.java
package com.yugutou.charpter11_bit.level2;

import java.util.Arrays;

public class CountBits {
    public static void main(String[] args) {
        System.out.println(Arrays.toString(countBits(2)));
        System.out.println(Arrays.toString(countBits2(2)));
    }

    /**
     * 方法1：统一移位统计
     *
     * @param num
     * @return
     */
    public static int[] countBits(int num) {
        int[] bits = new int[num + 1];
        for (int i = 0; i <= num; i++) {
            for (int j = 0; j < 32; j++) {
                bits[i] += (i >> j) & 1;
            }
        }
        return bits;
    }

    /**
     * 方法2：通过技巧x &= (x - 1);计算
     *
     * @param num
     * @return
     */
    public static int[] countBits2(int num) {
        int[] bits = new int[num + 1];
        for (int i = 0; i <= num; i++) {
            bits[i] = countOnes(i);
        }
        return bits;
    }

    public static int countOnes(int x) {
        int ones = 0;
        while (x > 0) {
            x &= (x - 1);
            ones++;
        }
        return ones;
    }

}
